File: cms-server/routes/content.js
/**
 * 页面内容读写路由
 *
 * GET  /api/content/:pageKey   —— 读取某页面 JSON（需登录）
 * PUT  /api/content/:pageKey   —— 更新某页面 JSON（需对应权限）
 *
 * pageKey 合法值：
 *   全局类：nav / footer / consultation（仅超级管理员可写）
 *   页面类：home / about / visa / saudi-visa / enterprise / transport / insurance / inspection
 */

const express = require('express');
const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');
const { requireAuth, requireSuperAdmin } = require('../middleware/auth');
const { audit } = require('../middleware/audit');

const router = express.Router();

const CONTENT_DIR = path.join(__dirname, '../../content/pages');
const GLOBAL_DIR  = path.join(__dirname, '../../content/global');
const DB_PATH     = path.join(__dirname, '../db/cms.db');

if (!fs.existsSync(CONTENT_DIR)) fs.mkdirSync(CONTENT_DIR, { recursive: true });
if (!fs.existsSync(GLOBAL_DIR))  fs.mkdirSync(GLOBAL_DIR,  { recursive: true });

const VALID_PAGE_KEYS = [
  'home', 'about', 'visa', 'saudi-visa',
  'enterprise', 'transport', 'insurance', 'inspection'
];

function getDB() {
  return new Database(DB_PATH);
}

// ── 权限检查辅助 ──────────────────────────────────────
function checkPagePerm(userId, pageKey) {
  if (pageKey === 'global' || ['nav','footer','consultation'].includes(pageKey)) return true; // 由调用方检查 super_admin
  const db = getDB();
  const row = db.prepare(
    'SELECT 1 FROM page_permissions WHERE user_id = ? AND page_key = ?'
  ).get(userId, pageKey);
  db.close();
  return !!row;
}

// ── GET /api/content/:pageKey ────────────────────────
// 无需认证（预览模式需要访问）
router.get('/:pageKey', (req, res) => {
  const { pageKey } = req.params;

  // 全局配置
  if (['nav','footer','consultation'].includes(pageKey)) {
    const filePath = path.join(GLOBAL_DIR, `${pageKey}.json`);
    return res.json(fs.existsSync(filePath) ? JSON.parse(fs.readFileSync(filePath, 'utf8')) : {});
  }

  if (!VALID_PAGE_KEYS.includes(pageKey)) {
    return res.status(400).json({ error: `无效的页面 key: ${pageKey}` });
  }

  const filePath = path.join(CONTENT_DIR, `${pageKey}.json`);
  res.json(fs.existsSync(filePath) ? JSON.parse(fs.readFileSync(filePath, 'utf8')) : {});
});

// ── PUT /api/content/:pageKey ────────────────────────
router.put('/:pageKey', requireAuth, (req, res) => {
  const { pageKey } = req.params;
  const user = req.user;
  const body = req.body;

  // 全局配置：仅超级管理员
  if (['nav','footer','consultation'].includes(pageKey)) {
    if (user.role !== 'super_admin') {
      return res.status(403).json({ error: '需要超级管理员权限' });
    }
    const filePath = path.join(GLOBAL_DIR, `${pageKey}.json`);
    fs.writeFileSync(filePath, JSON.stringify(body, null, 2), 'utf8');
    audit(req, 'update_global', pageKey, '');
    return res.json({ message: '全局配置已保存' });
  }

  if (!VALID_PAGE_KEYS.includes(pageKey)) {
    return res.status(400).json({ error: `无效的页面 key: ${pageKey}` });
  }

  // 普通编辑：检查页面权限
  if (user.role !== 'super_admin' && !checkPagePerm(user.id, pageKey)) {
    return res.status(403).json({ error: `无 [${pageKey}] 页面编辑权限` });
  }

  const filePath = path.join(CONTENT_DIR, `${pageKey}.json`);
  try {
    fs.writeFileSync(filePath, JSON.stringify(body, null, 2), 'utf8');
    audit(req, 'update_page', pageKey, '');
    res.json({ message: '页面内容已保存' });
  } catch (e) {
    res.status(500).json({ error: '写入失败：' + e.message });
  }
});

module.exports = router;

File: cms-server-go/routes/content.go
package routes

import (
	"cms-server-go/db"
	"cms-server-go/middleware"
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"

	"github.com/gin-gonic/gin"
)

var (
	contentDir string
	globalDir  string
)

// 合法的页面 key 列表
var validPageKeys = map[string]bool{
	"home": true, "about": true, "visa": true, "saudi-visa": true,
	"enterprise": true, "transport": true, "insurance": true, "inspection": true,
	"saudi-news": true,
}

// 全局配置 key
var globalKeys = map[string]bool{
	"nav": true, "footer": true, "consultation": true,
}

// InitContentDirs 初始化内容目录
func InitContentDirs(baseDir string) {
	contentDir = filepath.Join(baseDir, "content", "pages")
	globalDir = filepath.Join(baseDir, "content", "global")

	os.MkdirAll(contentDir, 0755)
	os.MkdirAll(globalDir, 0755)
}

// RegisterContentRoutes 注册内容路由
func RegisterContentRoutes(r *gin.RouterGroup) {
	content := r.Group("/content")
	{
		content.GET("/:pageKey", getContent)
		content.PUT("/:pageKey", middleware.RequireAuth(), updateContent)
	}
}

// getContent GET /api/content/:pageKey
func getContent(c *gin.Context) {
	pageKey := c.Param("pageKey")

	// 全局配置
	if globalKeys[pageKey] {
		filePath := filepath.Join(globalDir, pageKey+".json")
		data, err := os.ReadFile(filePath)
		if err != nil {
			c.JSON(http.StatusOK, gin.H{})
			return
		}
		var result interface{}
		if json.Unmarshal(data, &result) != nil {
			c.JSON(http.StatusOK, gin.H{})
			return
		}
		c.JSON(http.StatusOK, result)
		return
	}

	if !validPageKeys[pageKey] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的页面 key: " + pageKey})
		return
	}

	filePath := filepath.Join(contentDir, pageKey+".json")
	data, err := os.ReadFile(filePath)
	if err != nil {
		c.JSON(http.StatusOK, gin.H{})
		return
	}
	var result interface{}
	if json.Unmarshal(data, &result) != nil {
		c.JSON(http.StatusOK, gin.H{})
		return
	}
	c.JSON(http.StatusOK, result)
}

// updateContent PUT /api/content/:pageKey
func updateContent(c *gin.Context) {
	pageKey := c.Param("pageKey")
	role, _ := c.Get("role")
	userID, _ := c.Get("userID")

	var body interface{}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的 JSON 数据"})
		return
	}

	// 全局配置：仅超级管理员
	if globalKeys[pageKey] {
		if role != "super_admin" {
			c.JSON(http.StatusForbidden, gin.H{"error": "需要超级管理员权限"})
			return
		}
		filePath := filepath.Join(globalDir, pageKey+".json")
		data, _ := json.MarshalIndent(body, "", "  ")
		if err := os.WriteFile(filePath, data, 0644); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "写入失败：" + err.Error()})
			return
		}
		middleware.Audit(c, "update_global", pageKey, "")
		c.JSON(http.StatusOK, gin.H{"message": "全局配置已保存"})
		return
	}

	if !validPageKeys[pageKey] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的页面 key: " + pageKey})
		return
	}

	// 检查页面权限
	if role != "super_admin" {
		var count int
		err := db.DB.QueryRow(
			"SELECT COUNT(*) FROM page_permissions WHERE user_id = ? AND page_key = ?",
			userID, pageKey,
		).Scan(&count)
		if err != nil || count == 0 {
			c.JSON(http.StatusForbidden, gin.H{"error": "无 [" + pageKey + "] 页面编辑权限"})
			return
		}
	}

	filePath := filepath.Join(contentDir, pageKey+".json")
	data, _ := json.MarshalIndent(body, "", "  ")
	if err := os.WriteFile(filePath, data, 0644); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "写入失败：" + err.Error()})
		return
	}

	middleware.Audit(c, "update_page", pageKey, "")
	c.JSON(http.StatusOK, gin.H{"message": "页面内容已保存"})
}
